<!DOCTYPE html>
<html>
<head>
	<title>Neovim error: Channel doesn't exist</title>
	<link href="/style.css" rel="stylesheet">
	<meta charset="UTF-8" />
	<meta name="viewport" content="width=device-width, initial-scale=1.0" />
</head>
<body>
	<h1 id="title">
		<a href="/">
			Jakski's blog
		</a>
	</h1>
	<!-- START BODY -->
	<header>
		<h1 id="post-title">Neovim error: Channel doesn't exist</h1>
		<time id="post-date" datetime="2019-02-09">
			2019-02-09
		</time>
	</header>
	<p>After discovering Neovim's remote plugin interface I decided to give it a try
and develop some Python-hosted plugins. While I like the idea of moving plugin's
logic from Vim script to external language-agnostic services, I've already
stumbled upon not so trivial problems with debugging inter-process
communication.</p>
<h1>Scenario</h1>
<p>So I launched Neovim hoping to see Python host update syntax highlight in my
editor, but then this showed up:</p>
<pre><code>E475: Invalid argument: Channel doesn't exist
</code></pre>
<h1>Debbuging with documented methods</h1>
<p>Pynvim has pretty good documentation, including debugging
<a href="https://pynvim.readthedocs.io/en/latest/development.html">steps</a>. Let's see
what I can get after launching Neovim with debugging flags for Python-host:</p>
<pre><code>$ NVIM_PYTHON_LOG_FILE=logfile NVIM_PYTHON_LOG_LEVEL=DEBUG nvim
</code></pre>
<p>Now I can get trace from file <code>logfile_py3_rplugin</code> after error occurs:</p>
<pre><code>2019-02-09 20:33:30,425 [DEBUG @ msgpack_stream.py:_on_data:61] 31297 - received message: [2, b'nvim_error_event', [1, b'Message is not an array']]
...
...
...
2019-02-09 20:33:30,466 [ERROR @ base_events.py:default_exception_handler:1608] 31297 - Exception in callback _UnixWritePipeTransport._call_connection_lost(None)
handle: &lt;Handle _UnixWritePipeTransport._call_connection_lost(None)&gt;
BrokenPipeError: [Errno 32] Broken pipe

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
 File &quot;/usr/lib/python3.7/asyncio/events.py&quot;, line 88, in _run
   self._context.run(self._callback, *self._args)
 File &quot;/usr/lib/python3.7/asyncio/unix_events.py&quot;, line 744, in _call_connection_lost
   self._pipe.close()
BrokenPipeError: [Errno 32] Broken pipe
</code></pre>
<p>Ok, so Python-host dies for some reason and it seems it sends to Neovim some
malformed messages, but I can't find out what's being actually sent.</p>
<h1>Debugging by tracing system calls</h1>
<p>Before reading this part you must understand how Python-host gets started and
communicates with editor. Remote plugins take more resources than Vim scripts,
so Neovim tries to use them only when necessary. That's the reason you have to
run <em>:UpdateRemotePlugins</em>, before using any of them. <em>:UpdateRemotePlugins</em>
creates sort of a mapping between editor actions and remote procedure calls.
Every time user triggers RPC, Neovim starts plugin host if it's not yet running,
and forwards messages.</p>
<p>Plugin hosts are just Neovim's subprocesses communicating via standard
input/output. You can see it by triggering remote plugin and running <em>ps -ef
--forest</em>:</p>
<pre><code>jakub    20904  2164  0 13:47 pts/5    00:00:00  |   \_ -bash
jakub    29952 20904  0 19:53 pts/5    00:00:03  |   |   \_ vim
jakub    29989 29952  0 19:53 ?        00:00:01  |   |       \_ python3 -c import sys; sys.path.remove(&quot;&quot;); import neovim; neovim.start_host() /home/jakub/.vim/plugged/vim-yaml/rplugin/python3/vim_yaml
</code></pre>
<p>I'm going to track messages sent by process 29989. Every time plugin host
communicates with Neovim it issues system call to write something to file
descriptor 1(standard output). I can spy on messages with <em>strace</em>:</p>
<pre><code>$ strace -e write=1 -p 29989
strace: Process 31563 attached
epoll_wait(3,
</code></pre>
<p><em>strace</em> is really verbose, so I limited it to show only writes to standard
output. Now I have to reproduce error in editor. Just before error occurs I
see:</p>
<pre><code>write(1, &quot;StreamStartToken 0 0 0\nDocumentS&quot;..., 8195) = 8195
 | 00000  53 74 72 65 61 6d 53 74  61 72 74 54 6f 6b 65 6e  StreamStartToken |
 | 00010  20 30 20 30 20 30 0a 44  6f 63 75 6d 65 6e 74 53   0 0 0.DocumentS |
 | 00020  74 61 72 74 54 6f 6b 65  6e 20 30 20 30 20 33 0a  tartToken 0 0 3. |
 | 00030  42 6c 6f 63 6b 4d 61 70  70 69 6e 67 53 74 61 72  BlockMappingStar |
 | 00040  74 54 6f 6b 65 6e 20 32  20 30 20 30 0a 4b 65 79  tToken 2 0 0.Key |
...
...
...
</code></pre>
<p>This certainly doesn't look like RPC message sent by plugin host, so what's
that? It turned out to be one of my debug prints which worked fine, when I ran
tests with Tox, but caused headache for Neovim, because it's not proper msgpack
payload accepted as RPC. Removing single <em>print(...)</em> solved everything.</p>

	<!-- END BODY -->
</body>
</html>
